### PR TITLE
Fix bug with complement rule parsing in Object::procString

### PR DESCRIPTION
**Pull request by martyngigg**
Fix a bug in the processing of the MCNPX-style algebra strings that are used to create objects suitable for intersection calculations. 

**To test:**

This can only be confirmed by examining the `Object` created when a string is parsed. A unit test has been added that extensively checks the `Object`'s rule tree and surface collection. Examine the changes along with this test.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
The following images shows the system test differences are minor:

**LOQCentreNoGrav:**
loqcentrenograv

**SANS2DLOQReloadWorkspaces:**
sans2dloqreloadworkspaces

**Comment by martyngigg** [comment]
I've snuck a cppcheck fix that appeared with the new version and a recent merge to master.

**Comment by FedeMPouzols** [comment]
I'm having a look at this but as I'm not really familiar with `Geometry` if someone else more familiar with this pops in that would be good. So I won't take ownership for now. Otherwise I'll come back a bit later.

File: Framework/Geometry/test/ObjectTest.h [hunk 1]
```diff
@@ -392,7 +393,7 @@ class ObjectTest : public CxxTest::TestSuite {
     checkTrackIntercept(track, expectedResults);
   }
 
-  void xtestTrackTwoIsolatedCubes()
+  void testTrackTwoIsolatedCubes()
```
Comment (FedeMPouzols) [inline Framework/Geometry/test/ObjectTest.h]: Probably the x was just left there by accident.
This test was disabled but it seems to pass both before and after the changes in this branch.

Comment (martyngigg) [inline Framework/Geometry/test/ObjectTest.h]: Oh yes, I did inadvertently reactivated that test. If it passes then I may as well leave it as it is.

**Comment by NickDraper** [comment]
Given I've probably looked at this code as much as anyone, with the possible recent exception of Owen (and I don't know he has dived this deep), I've picked this up.

Code additions & removals look good.  Most importantly the additional tests prove that it works.

**Comment by NickDraper** [comment]
:shipit:

**Pull request merged**